[d2c LINK](https://dare2compete.com/o/pratyaksh-data-visualization-competition-coalesce-21-goa-institute-of-management-gim-goa-149186)



In [ ]:
#importing the libraries
import pandas as pd
import numpy as np

In [ ]:
##reading the csv file
file_agriculture = pd.read_csv("agriculture.csv")

In [ ]:
##coping the data
file_agri_copy = file_agriculture.copy()

In [ ]:
##checking for null values
file_agri_copy.isnull().sum()

Quarter            0
Encrypted Email    0
Country            0
Company Name       0
Tier               0
Employee Size      0
Industry           0
Revenue            0
Margin             0
Orders             0
AOV                0
Units              0
Recency            0
dtype: int64

In [ ]:
##checking info of the data
file_agri_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1603 entries, 0 to 1602
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Quarter          1603 non-null   object
 1   Encrypted Email  1603 non-null   object
 2   Country          1603 non-null   object
 3   Company Name     1603 non-null   object
 4   Tier             1603 non-null   object
 5   Employee Size    1603 non-null   object
 6   Industry         1603 non-null   object
 7   Revenue          1603 non-null   object
 8   Margin           1603 non-null   object
 9   Orders           1603 non-null   int64 
 10  AOV              1603 non-null   object
 11  Units            1603 non-null   int64 
 12  Recency          1603 non-null   int64 
dtypes: int64(3), object(10)
memory usage: 162.9+ KB


In [ ]:
file_agri_copy['Company Name'].unique()

array(['Customer211', 'Customer255', 'Customer262', ..., 'Customer14081',
       'Customer19668', 'Customer14276'], dtype=object)

In [ ]:
len(file_agri_copy['Company Name'].unique())

1558

unique = 1558, total companies = 1603
*therefore 45 cells / records mentioned belong to companies that already exists in the record*

In [ ]:
file_agri_copy['Country'].unique()

array(['United States', 'United Kingdom', 'Australia'], dtype=object)

In [ ]:
data = file_agri_copy.drop(columns = 'Employee Size')

In [ ]:
data.drop(["Tier","AOV",],axis = 1,inplace = True)

In [ ]:
data

Quarter         Country   Company Name  ... Orders Units  Recency
0     FY19'Q1   United States    Customer211  ...      1     2      986
1     FY19'Q1   United States    Customer255  ...      1     3      972
2     FY19'Q1   United States    Customer262  ...      1     1     1011
3     FY19'Q1   United States    Customer263  ...      1     1     1022
4     FY19'Q1   United States    Customer317  ...      1     2      992
...       ...             ...            ...  ...    ...   ...      ...
1598  FY20'Q2   United States  Customer19346  ...      1     4      556
1599  FY20'Q2  United Kingdom   Customer6930  ...      2    14      540
1600  FY20'Q2  United Kingdom  Customer14081  ...      2     2      572
1601  FY20'Q2  United Kingdom  Customer19668  ...      1     1      544
1602  FY20'Q2   United States  Customer14276  ...      5     6      547

[1603 rows x 8 columns]

In [ ]:
data.groupby('Company Name').sum()
##data would be collective irrespective of the quarter 1,2,3,4 of '19 and 1,2 of 2020.Aggregate ke liye

Orders  Units  Recency
Company Name                         
0                   1      1      997
Customer1001        2      5      987
Customer1003        1      5      972
Customer10052       1      1      682
Customer1006        1      1     1046
...               ...    ...      ...
Customer997         1      4     1046
Customer9971        3      3      727
Customer9977        1      1      730
Customer998         1      1      976
Customer9982        1      1      696

[1558 rows x 3 columns]

##Next field : *Communication* 
using the sales_sheet.csv this time

In [ ]:
file_2 = pd.read_csv("sales_sheet.csv")

In [ ]:
file_2

Quarter                   Encrypted Email  ... Units Recency
0      FY19'Q1  757043b476800d67ca6b793fe22610b3  ...     3    1019
1      FY19'Q1  815c656d552cbc56ecab9304ad17540c  ...     1    1001
2      FY19'Q1  ae69b698191be39cf0a60fd6e16165f5  ...     3     962
3      FY19'Q1  aea54c4b6e13731744b89a5873e262c5  ...    16    1005
4      FY19'Q1  d6e2891285b1ff2dbe02b44d2ef8a8c1  ...     4    1036
...        ...                               ...  ...   ...     ...
29104  FY20'Q2  24ed3285cb100aa6bfd0fd67c19d3b8c  ...     1     529
29105  FY20'Q2  151607a90c2400fdeaa5d24d626cedd4  ...     1     513
29106  FY20'Q2  8989c5021b793cb33c598f3cb58523ad  ...     1     535
29107  FY20'Q2  3407ea6d2620079eb0f13e96ed1eaebe  ...     1     548
29108  FY20'Q2  a1108bba82eb3566958bb9303f85ef07  ...     1     508

[29109 rows x 13 columns]

In [ ]:
len(file_2['Company Name'].unique())

20026

# Repeating entries = 9088
*unique company entries = 20026;
Total company entries = 29109*


In [ ]:
## Descibing the data
file_2.describe()

Orders         Units       Recency
count  29109.000000  29109.000000  29109.000000
mean       1.595829      4.724518    706.122849
std        1.866876     10.976201    146.732250
min        1.000000      1.000000    504.000000
25%        1.000000      1.000000    577.000000
50%        1.000000      2.000000    677.000000
75%        2.000000      4.000000    819.000000
max       98.000000    422.000000   1051.000000

In [ ]:
##Let's remove the irrelevant columns to make ouor work easier. uske pehle thoda covariance nikalte hai just yuhi aise hi.
print(file_2.corr(method='pearson'))

           Orders     Units   Recency
Orders   1.000000  0.471663 -0.020302
Units    0.471663  1.000000  0.013554
Recency -0.020302  0.013554  1.000000


In [ ]:
print(file_2.corr(method='kendall'))

           Orders     Units   Recency
Orders   1.000000  0.550252 -0.009351
Units    0.550252  1.000000  0.019686
Recency -0.009351  0.019686  1.000000


In [ ]:
print(file_2.corr(method='spearman'))

           Orders    Units   Recency
Orders   1.000000  0.61454 -0.011846
Units    0.614540  1.00000  0.026800
Recency -0.011846  0.02680  1.000000


In [ ]:
##most occurences
file_2['Quarter'].mode()
file_2['Quarter'].value_counts()
##the companies have been retained + increased as we see

FY20'Q2    8538
FY20'Q1    6582
FY19'Q4    4840
FY19'Q3    4029
FY19'Q2    3123
FY19'Q1    1997
Name: Quarter, dtype: int64

In [ ]:
file_2['Industry'].mode()
file_2['Industry'].value_counts()

Communication                        5753
Manufacturing & Natural Resources    4456
Heath & Welfare                      3060
Transportation                       2570
Financial                            2368
Education                            2316
 Retail & Wholesale                  2144
Agriculture                          1603
Construction                         1473
Services                              979
Retail & Wholesale                    769
Nonprofit and Charity                 762
Utilities                             451
Government                            405
Name: Industry, dtype: int64

In [ ]:
file_2['Tier'].mode()
file_2['Tier'].value_counts()
##convert base to silvers
##upgrading the silvers to golds
##maintaining the gold

Base      26520
Silver     2346
Gold        243
Name: Tier, dtype: int64

In [ ]:
file_2['Company Name'].mode()
file_2['Company Name'].value_counts()

Customer4525     30
Customer617      22
Customer6262     19
Customer2087     17
Customer3169     16
                 ..
Customer13976     1
Customer16049     1
Customer12677     1
Customer7025      1
Customer5379      1
Name: Company Name, Length: 20026, dtype: int64

In [ ]:
file_2['Employee Size'].mode()
file_2['Employee Size'].value_counts()

1-9 employees          7244
20-49 employees        6860
50-99 employees        4920
10-19 employees        3452
100-499 employees      2767
100-299 employees      1788
500-999 employees       678
100+ employees          401
300-499 employees       385
1200+ employees         231
1000-2999 employees     174
3000+ employees         112
1000-1199 employees      50
500+ employees           43
1000+ employees           4
Name: Employee Size, dtype: int64

In [ ]:
file_2['Country'].mode()
file_2['Country'].value_counts()
##we need acquisitions in UK

United States     16455
United Kingdom    10834
Australia          1820
Name: Country, dtype: int64

In [ ]:
##making a copy of the file of sales_sheet to edit the dataframe
file_3 = file_2.copy()

In [ ]:
file_3.drop(["Tier","AOV","Employee Size","Encrypted Email"],axis = 1,inplace = True)

In [ ]:
file_3

Quarter         Country   Company Name  ... Orders Units Recency
0      FY19'Q1       Australia      Customer1  ...      1     3    1019
1      FY19'Q1  United Kingdom      Customer2  ...      1     1    1001
2      FY19'Q1   United States      Customer3  ...      1     3     962
3      FY19'Q1   United States      Customer4  ...      2    16    1005
4      FY19'Q1   United States      Customer5  ...      1     4    1036
...        ...             ...            ...  ...    ...   ...     ...
29104  FY20'Q2  United Kingdom  Customer20022  ...      1     1     529
29105  FY20'Q2  United Kingdom  Customer20023  ...      1     1     513
29106  FY20'Q2  United Kingdom  Customer20024  ...      1     1     535
29107  FY20'Q2  United Kingdom  Customer20025  ...      1     1     548
29108  FY20'Q2  United Kingdom  Customer20026  ...      1     1     508

[29109 rows x 9 columns]

In [ ]:
file_4 = file_3.copy()

In [ ]:
##for aggregate data
file_4.groupby('Company Name').sum()
##recency getting added too.we can remove the recency column

Orders  Units  Recency
Company Name                        
0                  1      1      997
Customer1          1      3     1019
Customer10         5      8     1919
Customer100       11     13     3972
Customer1000       1     10      971
...              ...    ...      ...
Customer9995       2      6     1275
Customer9996       4      4     1388
Customer9997       2     11      718
Customer9998       5     23     1995
Customer9999       1      1      731

[20026 rows x 3 columns]

In [ ]:
file_4

Quarter         Country   Company Name  ... Orders Units Recency
0      FY19'Q1       Australia      Customer1  ...      1     3    1019
1      FY19'Q1  United Kingdom      Customer2  ...      1     1    1001
2      FY19'Q1   United States      Customer3  ...      1     3     962
3      FY19'Q1   United States      Customer4  ...      2    16    1005
4      FY19'Q1   United States      Customer5  ...      1     4    1036
...        ...             ...            ...  ...    ...   ...     ...
29104  FY20'Q2  United Kingdom  Customer20022  ...      1     1     529
29105  FY20'Q2  United Kingdom  Customer20023  ...      1     1     513
29106  FY20'Q2  United Kingdom  Customer20024  ...      1     1     535
29107  FY20'Q2  United Kingdom  Customer20025  ...      1     1     548
29108  FY20'Q2  United Kingdom  Customer20026  ...      1     1     508

[29109 rows x 9 columns]

In [ ]:
##need to convert revenue and margins to int values : late yaad aaya :(, iska matlab sab phirse karna 

# ***Here We Go Again***

In [71]:
file_final = pd.read_csv('sales_sheet_1.csv')

In [72]:
df_1 = file_final.copy()

In [73]:
df_1.describe()

Revenue        Orders         Units       Recency
count   29109.000000  29109.000000  29109.000000  29109.000000
mean     4698.240819      1.595829      4.724518    706.122849
std     10051.533506      1.866876     10.976201    146.732250
min        14.000000      1.000000      1.000000    504.000000
25%      1124.000000      1.000000      1.000000    577.000000
50%      1869.000000      1.000000      2.000000    677.000000
75%      4209.000000      2.000000      4.000000    819.000000
max    298882.000000     98.000000    422.000000   1051.000000

In [78]:
df_1.drop(["Tier","AOV","Employee Size","Encrypted Email"],axis = 1,inplace = True)

KeyError: ignored

In [75]:
df_1

Quarter         Country   Company Name  ... Orders  Units Recency
0      FY19'Q1       Australia      Customer1  ...      1      3    1019
1      FY19'Q1  United Kingdom      Customer2  ...      1      1    1001
2      FY19'Q1   United States      Customer3  ...      1      3     962
3      FY19'Q1   United States      Customer4  ...      2     16    1005
4      FY19'Q1   United States      Customer5  ...      1      4    1036
...        ...             ...            ...  ...    ...    ...     ...
29104  FY20'Q2  United Kingdom  Customer20022  ...      1      1     529
29105  FY20'Q2  United Kingdom  Customer20023  ...      1      1     513
29106  FY20'Q2  United Kingdom  Customer20024  ...      1      1     535
29107  FY20'Q2  United Kingdom  Customer20025  ...      1      1     548
29108  FY20'Q2  United Kingdom  Customer20026  ...      1      1     508

[29109 rows x 9 columns]

In [80]:
df_2 = df_1.copy()

In [84]:
df_2.groupby(['Company Name']).sum()

Revenue  Orders  Units  Recency
Company Name                                 
0                 458       1      1      997
Customer1        1586       1      3     1019
Customer10       4415       5      8     1919
Customer100     17263      11     13     3972
Customer1000     6832       1     10      971
...               ...     ...    ...      ...
Customer9995     3348       2      6     1275
Customer9996     5024       4      4     1388
Customer9997     5724       2     11      718
Customer9998    12292       5     23     1995
Customer9999      549       1      1      731

[20026 rows x 4 columns]

In [92]:
##downloading csv file
df_2.to_csv(r'C:\Users\Aaditee\Desktop\projects\P\processed_data_for_Q5.csv', index=False)

In [93]:
checking = pd.read_csv('Q5.csv')

In [94]:
checking

Quarter         Country   Company Name  ... Margin  Orders Units
0      FY19'Q1       Australia      Customer1  ...     47       1     3
1      FY19'Q1  United Kingdom      Customer2  ...      0       1     1
2      FY19'Q1   United States      Customer3  ...      0       1     3
3      FY19'Q1   United States      Customer4  ...    930       2    16
4      FY19'Q1   United States      Customer5  ...    150       1     4
...        ...             ...            ...  ...    ...     ...   ...
29104  FY20'Q2  United Kingdom  Customer20022  ...     61       1     1
29105  FY20'Q2  United Kingdom  Customer20023  ...      0       1     1
29106  FY20'Q2  United Kingdom  Customer20024  ...    175       1     1
29107  FY20'Q2  United Kingdom  Customer20025  ...     34       1     1
29108  FY20'Q2  United Kingdom  Customer20026  ...     54       1     1

[29109 rows x 8 columns]

In [95]:
len(checking['Company Name'].unique())

20026

In [96]:
df_3 = df_2.groupby(['Company Name']).sum()

In [98]:
df_3

Revenue  Orders  Units  Recency
Company Name                                 
0                 458       1      1      997
Customer1        1586       1      3     1019
Customer10       4415       5      8     1919
Customer100     17263      11     13     3972
Customer1000     6832       1     10      971
...               ...     ...    ...      ...
Customer9995     3348       2      6     1275
Customer9996     5024       4      4     1388
Customer9997     5724       2     11      718
Customer9998    12292       5     23     1995
Customer9999      549       1      1      731

[20026 rows x 4 columns]

In [99]:
df_3.to_csv(r'processed_data_for_Q5.csv', index=False)

In [ ]:
##yeh data barbaad hai. plis go to dusra file